lib/passport-wsfed-saml2/samlp.js

```diff
@@ -286,80 +301,122 @@ Samlp.prototype = {
 
     samlResponse = utils.parseSamlResponse(samlResponse);
 
-    // check status
-    var samlStatus = self.getSamlStatus(samlResponse);
-
-    // Check if this is a known error
-    var errorMessage = ErrorMessages[samlStatus.subCode] ||
-                       ErrorMessages[samlStatus.code];
-
-    if (errorMessage) {
-      // Return auth failed with the actual message or a friendly message
-      return callback (new AuthenticationFailedError(samlStatus.message || errorMessage, samlStatus.detail));
+    // Check that the saml Resopnse actually has a Response object
+    var responseXML = xpath.select("//*[local-name(.)='Response']", samlResponse)[0];
+    if (!responseXML){ 
+      return callback(new Error('XML is not a valid saml response'));
     }
 
-    // extract assertion
-    self.extractAssertion(samlResponse, function (err, assertion) {
-      if (err) { return callback(err); }
-      if (!assertion) {
-        return callback(new Error('saml response does not contain an Assertion element (Status: ' + samlStatus.code + ')'));
-      }
+    self.isValidResponseID(responseXML.getAttribute('ID'), function(err){
+      if (err) return callback(err);
 
-      var samlResponseSignaturePath = "//*[local-name(.)='Response']/*[local-name(.)='Signature' and namespace-uri(.)='http://www.w3.org/2000/09/xmldsig#']";
-      var isResponseSigned = xpath.select(samlResponseSignaturePath, samlResponse).length > 0;
-      var samlAssertionSignaturePath = "//*[local-name(.)='Assertion']/*[local-name(.)='Signature' and namespace-uri(.)='http://www.w3.org/2000/09/xmldsig#']";
-      var isAssertionSigned =  xpath.select(samlAssertionSignaturePath, assertion).length > 0;
+      var inResponseTo = responseXML.getAttribute('InResponseTo');
+
+      self.isValidInResponseTo(inResponseTo, function(err){
```

Comment:
It would be nice to move this to an async waterfall/series to make it more redeadable